shopinvader_product_stock_forecast/models/shopinvader_variant.py

```diff
@@ -0,0 +1,99 @@
+                WHEN whd.id IS NOT NULL AND whs.id IS NULL THEN m.product_qty
+                END
+            ) AS qty
+        """
+
+    def _prepare_stock_forecast_from(self):
+        return """
+            stock_move m
+            LEFT JOIN stock_location ls ON (ls.id=m.location_id)
+            LEFT JOIN stock_location ld ON (ld.id=m.location_dest_id)
+            LEFT JOIN stock_warehouse whs
+                ON ls.parent_path LIKE CONCAT('%%/', whs.view_location_id, '/%%')
+            LEFT JOIN stock_warehouse whd
+                ON ld.parent_path LIKE CONCAT('%%/', whd.view_location_id, '/%%')
+        """
+
+    def _prepare_stock_forecast_group_by(self):
+        return "m.date"
+
+    def _prepare_stock_forecast_where_clause(self):
+        """Prepare the where and where params for the forecast query"""
+        where_clause = """
+            m.product_id = %s
+            AND m.state NOT IN ('cancel', 'draft', 'done')
+            AND (whs.id IS NOT NULL OR whd.id IS NOT NULL)
+            AND (whs.id IS NULL OR whd.id IS NULL OR whs.id != whd.id)
+        """
+        where_clause_params = [self.record_id.id]
+        # Warehouse
+        warehouse_ids = self.env.context.get("warehouse", [])
+        if warehouse_ids:
+            where_clause += """
+                AND (
+                    CASE
+                    WHEN whs.id IS NOT NULL AND whd.id IS NULL THEN whs.id
+                    WHEN whd.id IS NOT NULL AND whs.id IS NULL THEN whd.id
+                    END
+                ) IN %s
+            """
+            where_clause_params.append(tuple(warehouse_ids))
+        # Forecast Horizon
+        if self.backend_id.product_stock_forecast_horizon:
+            horizon = self.backend_id.product_stock_forecast_horizon
+            dt_end = fields.Datetime.now() + timedelta(days=horizon)
+            where_clause += "\nAND m.date <= %s"
+            where_clause_params.append(dt_end)
+        return where_clause, where_clause_params
+
+    def _prepare_stock_forecast_data(self):
+        """Prepare the stock forecast data"""
+        query = sql.SQL(
+            """
+            SELECT {select}
+            FROM {from_clause}
+            WHERE {where_clause}
+            GROUP BY {group_by}
+            """
+        )
+        where_clause, params = self._prepare_stock_forecast_where_clause()
+        query = query.format(
+            select=sql.SQL(self._prepare_stock_forecast_select()),
+            from_clause=sql.SQL(self._prepare_stock_forecast_from()),
+            where_clause=sql.SQL(where_clause),
+            group_by=sql.SQL(self._prepare_stock_forecast_group_by()),
+        )
+        self.env["base"].flush()
+        self.env.cr.execute(query, params)
+        data = self.env.cr.dictfetchall()
+        # Convert date to string, for it to be serializable
+        for row in data:
+            row["date"] = fields.Datetime.to_string(row["date"])
```

Comment:
we should probably use this kind of encoder  https://github.com/OCA/rest-framework/blob/14.0/base_rest/http.py#L47 everywhere